cmd/gapit/replace_resource.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (C) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+
+	"github.com/google/gapid/core/app"
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/service"
+)
+
+type replaceResourceVerb struct{ ReplaceResourceFlags }
+
+func init() {
+	verb := &replaceResourceVerb{
+		ReplaceResourceFlags{
+			At:              -1,
+			OutputTraceFile: "newcapture.gfxtrace",
+		},
+	}
+	app.AddVerb(&app.Verb{
+		Name:      "replace_resource",
+		ShortHelp: "Produce a new trace with the given resource replaced at the given command",
+		Action:    verb,
+	})
+}
+
+func (verb *replaceResourceVerb) Run(ctx context.Context, flags flag.FlagSet) error {
+	if flags.NArg() != 3 {
+		app.Usage(ctx, "Exactly 3 arguments: resource handle, resource path, capture, got %d", flags.NArg())
+		return nil
+	}
+	resourceHandle := flags.Arg(0)
+	newResourceFilename := flags.Arg(1)
+
+	captureFilepath, err := filepath.Abs(flags.Arg(2))
+	if err != nil {
+		return log.Errf(ctx, err, "Could not find capture file '%s'", flags.Arg(2))
+	}
+
+	client, err := getGapis(ctx, verb.Gapis, verb.Gapir)
+	if err != nil {
+		return log.Err(ctx, err, "Failed to connect to the GAPIS server")
+	}
+	defer client.Close()
+
+	capture, err := client.LoadCapture(ctx, captureFilepath)
+	if err != nil {
+		return log.Errf(ctx, err, "Failed to load the capture file '%v'", captureFilepath)
+	}
+
+	boxedResources, err := client.Get(ctx, capture.Resources().Path())
+	if err != nil {
+		return log.Err(ctx, err, "Could not find the capture's resources")
+	}
+	resources := boxedResources.(*service.Resources)
+
+	if verb.At == -1 {
+		boxedCapture, err := client.Get(ctx, capture.Path())
+		if err != nil {
+			return log.Err(ctx, err, "Failed to load the capture")
+		}
+		verb.At = int(boxedCapture.(*service.Capture).NumCommands) - 1
+	}
+
+	handleRegex := regexp.MustCompile("<(.*?)>")
+
+	for _, types := range resources.GetTypes() {
+		if types.Type == api.ResourceType_ShaderResource {
+			for _, v := range types.GetResources() {
+				handleHexStr := handleRegex.FindStringSubmatch(v.GetHandle())[1]
```

Comment:
If we can't match the handle in the "<*>" form, we should probably test against the full returned handle.

I think handles MAY come back with different names if the user has debug names attached to the handles, so matching against a "real" name may be helpful.

Reply:
I agree, making any sort of assumptions on the format of the handle string here seems fragile.
If the typical usage case is purely to find a shader by handle in hex, perhaps we can just make this a substring match (which would also then work for debug annotated names), and error if you find multiple of the same substring match?

Reply:
Done